src/parse.rs

```diff
@@ -321,6 +321,69 @@ pub fn parse_mailbox(
     }
 }
 
+pub fn parse_status(
+    mut lines: &[u8],
+    mailbox_name: &str,
+    unsolicited: &mut mpsc::Sender<UnsolicitedResponse>,
+) -> Result<Mailbox> {
+    let mut mailbox_opt = None;
+
+    loop {
+        if lines.is_empty() {
+            break mailbox_opt.ok_or_else(|| {
+                ParseError::Unexpected("did not receive STATUS response".to_string()).into()
+            });
+        }
+        match imap_proto::parser::parse_response(lines) {
+            Ok((
+                rest,
+                Response::MailboxData(MailboxDatum::Status {
+                    mailbox: their_mailbox_name,
+                    status,
+                }),
+            )) => {
+                lines = rest;
+                if mailbox_name != their_mailbox_name {
+                    return Err(ParseError::Unexpected(format!(
+                        "received STATUS for mailbox '{}', expected '{}'",
+                        their_mailbox_name, mailbox_name
+                    ))
+                    .into());
+                }
+                if mailbox_opt.is_some() {
+                    return Err(ParseError::Unexpected(format!(
+                        "received multiple STATUS responses for the same mailbox '{}'",
+                        mailbox_name
+                    ))
+                    .into());
+                }
+                let mut mailbox = Mailbox::default();
+                for attr in status {
+                    match attr {
+                        StatusAttribute::HighestModSeq(v) => mailbox.highest_mod_seq = Some(v),
+                        StatusAttribute::Messages(v) => mailbox.exists = v,
+                        StatusAttribute::Recent(v) => mailbox.recent = v,
+                        StatusAttribute::UidNext(v) => mailbox.uid_next = Some(v),
+                        StatusAttribute::UidValidity(v) => mailbox.uid_validity = Some(v),
+                        StatusAttribute::Unseen(v) => mailbox.unseen = Some(v),
+                        _ => {} // needed because StatusAttribute is #[non_exhaustive]
+                    }
+                }
+                mailbox_opt = Some(mailbox);
+            }
+            Ok((rest, data)) => {
+                lines = rest;
+                if let Some(resp) = try_handle_unilateral(data, unsolicited) {
+                    break Err(resp.into());
```

Comment:
There is a sort of weird case where we enter this case and `mailbox_opt.is_some()`. I'm not sure if the right thing to do in that instance is to return an `Err` or just ignore the spurious response and give the `Ok`. What do you think?

Reply:
Well, even if `mailbox_opt` is `None` (i.e. we haven't gotten the expected `STATUS` response yet), that response could still show up later in `lines`.  In particular, if the unknown response is some sort of unsolicited data sent on the server's initiative and unrelated to the `STATUS` command, then the ordering between it and the `STATUS` response would be a matter of chance.

It might be a good idea to make unknown untagged responses non-fatal in all contexts.  The IMAP spec does say: "A client MUST be prepared to accept any server response at all times.  This includes server data that was not requested."  This wording doesn't explicitly differentiate between known response types (that simply don't correspond to requests) and unknown response types, but arguably it could be read to cover both.

But I don't think it makes sense to make them non-fatal only in this case.

Reply:
I think it's probably okay to return it as an error, because the caller can always then decide what to do next. If they don't want to treat "weird" server messages as errors, they have that option. It only really gets weird in the case where there is _also_ successful return data for the command, since then all the caller can do is call the operation again. But that doesn't work if the operation isn't idempotent.

Honestly not sure what the right thing to do here is, but I agree with you that it's broader than the scope of this PR, so resolving.

cc @mordak if you want to think about what the right thing to do here is too.